pkg/acirenderer/acirenderer.go

```diff
@@ -0,0 +1,209 @@
+type RenderedACI []*ACIFiles
+
+// Return the RenderedACI list. All file outside rootfs are excluded (at the
+// moment only "manifest").
+func GetRenderedACIFromList(imgs Images, ap ACIProvider) (RenderedACI, error) {
+	if len(imgs) == 0 {
+		return nil, fmt.Errorf("image list empty")
+	}
+
+	allFiles := make(map[string]struct{})
+
+	renderedACI := RenderedACI{}
+	// This implementation needs to start from the end of the tree.
+	end := len(imgs) - 1
+
+	for i := 0; i <= end; i++ {
+		img := imgs[i]
+		pwlm := getUpperPWLM(imgs, img.level)
+		ra, err := getACIFiles(img, ap, allFiles, pwlm)
+		if err != nil {
+			return nil, err
+		}
+		renderedACI = append(renderedACI, ra)
+	}
+
+	return renderedACI, nil
+}
+
+func getUpperPWLM(imgs Images, level uint16) map[string]struct{} {
+	end := len(imgs) - 1
+	var pwlm map[string]struct{}
+	// The pwl to use is the first available starting from the upper level
+	curlevel := level
+	for i := 0; i <= end; i++ {
+		img := imgs[i]
+		if img.level < curlevel && len(img.im.PathWhitelist) > 0 {
+			pwlm = pwlToMap(img.im.PathWhitelist)
+			curlevel = level
+		}
+	}
+	return pwlm
+}
+
+// Return the ACIFiles struct for the given image. All file outside rootfs are
+// excluded (at the moment only "manifest").
+func getACIFiles(img Image, ap ACIProvider, allFiles map[string]struct{}, pwlm map[string]struct{}) (*ACIFiles, error) {
+	rs, err := ap.ReadStream(img.key)
+	if err != nil {
+		return nil, err
+	}
+	defer rs.Close()
+
+	hash := sha512.New()
+	r := io.TeeReader(rs, hash)
+
+	ra := &ACIFiles{fileMap: make(map[string]struct{})}
+	if err = Walk(tar.NewReader(r), func(hdr *tar.Header) error {
+		name := hdr.Name
+		cleanName := filepath.Clean(name)
+
+		// Ignore files outside /rootfs/ (at the moment only "manifest")
+		if !strings.HasPrefix(cleanName, "rootfs/") {
+			return nil
+		}
+
+		// Is the file in our PathWhiteList?
+		if len(img.im.PathWhitelist) > 0 {
+			thispwlm := pwlToMap(img.im.PathWhitelist)
+			if _, ok := thispwlm[cleanName]; !ok {
+				return nil
+			}
+		}
+		// Is the file in the previous level PathWhiteList?
+		if pwlm != nil {
+			if _, ok := pwlm[cleanName]; !ok {
+				return nil
+			}
+		}
+		// Is the file already provided by a previous image?
+		if _, ok := allFiles[cleanName]; ok {
+			return nil
+		}
+		ra.fileMap[name] = struct{}{}
+		allFiles[cleanName] = struct{}{}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	// Tar does not necessarily read the complete file, so ensure we read the entirety into the hash
+	if _, err := io.Copy(ioutil.Discard, r); err != nil {
+		return nil, fmt.Errorf("error reading ACI: %v", err)
+	}
+
+	if g := ap.HashToKey(hash); g != img.key {
+		return nil, fmt.Errorf("image hash does not match expected (%s != %s)", g, img.key)
+	}
+
+	ra.key = img.key
+	return ra, nil
+}
+
+// Convert pathWhiteList slice to a map for faster search
+// Also change path to be relative to "/" so it can easyly used without the
```

Comment:
> Also, you mean relative to "rootfs/", right?

Also this was related to the previous implementation. I tried to explain it better.